File: Binary Search/69. Sqrt(x)/Solution.py
class Solution:
    def mySqrt(self, x: int) -> int:
        # N : x
        # Time Complexity: O(logN)
        # Space Complexity: O(1)
        n = x
        l, r = 0, n
        while l <= r:
            mid = (r + l) // 2
            if x < mid * mid:
                r = mid - 1
            elif x > mid * mid:
                l = mid + 1
            else:
                return int(mid)
        return int(l) - 1

File: Binary Search/69. Sqrt(x)/Solution.java
class Solution {
    public int mySqrt(int x) {
        // N : x
        // Time Complexity: O(logN)
        // Space Complexity: O(1)
        long l = 0;
        long r = x;
        while (l <= r) {
            long mid = (r + l) / 2;
            if (x > mid * mid) l = mid + 1;
            else if (x < mid * mid) r = mid - 1;
            else return (int) mid;
        }
        return (int) l - 1;
    }
}